### ISSUE TITLE
PHP sessions block other requests when downloading protected files

### ISSUE THREAD

**Issue by lozcalver** [silverstripe/silverstripe-assets#232]
`ProtectedFileController` doesn’t close the current PHP session before serving the response, which means it’s not possible to download two protected files at the same time, or even browse the website while a protected file is downloading.

Adjusting [`ProtectedFileController::handleFile()`](https://github.com/silverstripe/silverstripe-assets/blob/1.3.3/src/Storage/ProtectedFileController.php#L65) to the following does fix the problem:

```php
$response = $this->getRouteHandler()->getResponseFor($filename);
session_write_close();
return $response;
```

Our current `Session` API doesn’t offer anything to close the session, perhaps that should be introduced as part of the fix for this...

**Comment by kinglozzer** [comment]
Would anyone object to just calling `session_write_close()` directly (i.e. not introducing it to the `Session` API)? AFAICT that should be safe enough

**Comment by chillu** [comment]
@kinglozzer That sounds good to me, wanna do a quick PR? Is there anything controversial about adding this to the `Session` API? I'm worried this might break https://github.com/silverstripe/silverstripe-hybridsessions or https://github.com/silverstripe/silverstripe-dynamodb if we add it directly - although I think they're both using custom PHP session handlers, which presumably also respect `session_write_close()`...

**Comment by kinglozzer** [comment]
> Is there anything controversial about adding this to the `Session` API?

I don’t think so, it just makes it easier to include in a patch release if we don’t add it to the `Session` API. If we do, we need to update the `silverstripe/framework` Composer constraint in this module. If we wanted this change in the `1.2` branch (and `4.2` of framework), I’m not even sure what the correct constraint would be... `"silverstripe/framework": "^4.2.6|^4.3.5|^4.4.4"`?

AFAICT hybridsessions/dynamodb both call `session_set_save_handler()` rather than interfere with `SilverStripe\Control\Session` in any way, so they _should_ be fine with that change...

**Comment by chillu** [comment]
@kinglozzer I think the dependency bump indicates to me that this should go into a minor release, not a patch release. So `assets:1.5` and `framework:4.5`. That makes it a bit easier than a patch level bugfix release of `assets` which then requires complex constraints on `framework`?

**Comment by GuySartorelli** [comment]
Closing in favour of https://github.com/silverstripe/silverstripe-framework/issues/11671
